### PR TITLE
ReduceLayer, allow reduce_mean of time dimension without flattening

### PR DESCRIPTION
**Pull request by vieting**
A time mask is applied to take different sequence lengths into account, the elements are divided by the sequence length and then tf.reduce_sum is applied. Flattening is removed.

### CODE REVIEW

File: tests/test_TFNetworkLayer.py [hunk 1]
```diff
@@ -3065,6 +3065,71 @@ def mask_func(source, **kwargs):
...
 
+def test_reduce_mean_in_time():
+  with make_scope() as session:
+    import numpy as np
+    n_out = 5
+    config = Config({
+      "debug_print_layer_output_template": True,
+      "extern_data": {
+        "data": {"dim": n_out},
+        "classes": {"dim": n_out},
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: You don't need classes. Do not add it then.

File: tests/test_TFNetworkLayer.py [hunk 2]
```diff
@@ -3065,6 +3065,71 @@ def mask_func(source, **kwargs):
       last_var_v = var_v
 
 
+def test_reduce_mean_in_time():
+  with make_scope() as session:
+    import numpy as np
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: Just use `numpy` directly. No need for the `np` alias. And `numpy` is already imported.

File: tests/test_TFNetworkLayer.py [hunk 3]
```diff
@@ -3065,6 +3065,71 @@ def mask_func(source, **kwargs):
...
+    config = Config({
+      "debug_print_layer_output_template": True,
+      "extern_data": {
+        "data": {"dim": n_out},
+        "classes": {"dim": n_out},
+      }})
+    net = TFNetwork(config=config, train_flag=True)
+    net.construct_from_dict({
+      "reduce_mean": {"class": "reduce", "mode": "mean", "axis": "T", "keep_dims": True, "from": ["data"],
+                      "is_output_layer": True}
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: If you call it `"output"` directly, you would not need `"is_output_layer": True`. But as you like.

File: tests/test_TFNetworkLayer.py [hunk 4]
```diff
@@ -3065,6 +3065,71 @@ def mask_func(source, **kwargs):
...
+    })
+    session.run(tf.global_variables_initializer())
+    out = net.layers["reduce_mean"].output.placeholder
+    n_batch = 3
+    max_seq_len = 10
+    feed = make_feed_dict(net.extern_data.data.values(), n_batch=n_batch, n_time=max_seq_len, same_time=True)
+    v = session.run(out, feed_dict=feed)
+    input_len = feed[list(feed.keys())[0]]
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: This is not a good idea. The order in `feed.keys()` is non-deterministic (if you have more than 1 entry). But anyway, you assume it is `"data"`, right? So just make it more explicit, sth like `feed[net.extern_data.data["data"].size_placeholder[0]]`.

File: tests/test_TFNetworkLayer.py [hunk 5]
```diff
@@ -3065,6 +3065,71 @@ def mask_func(source, **kwargs):
...
+    input_data = feed[list(feed.keys())[1]]
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: Same here.

File: tests/test_TFNetworkLayer.py [hunk 6]
```diff
@@ -3065,6 +3065,71 @@ def mask_func(source, **kwargs):
...
+
+    ref = np.zeros([n_batch, n_out])
+    for batch, len in enumerate(input_len):
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: Do not overwrite Python builtins (`len`). (PyCharm should warn about that, or not?)

File: tests/test_TFNetworkLayer.py [hunk 7]
```diff
@@ -3065,6 +3065,71 @@ def mask_func(source, **kwargs):
...
+      ref[batch, :] = np.mean(input_data[batch, :len, :], axis=0)
+    ref = np.expand_dims(ref, axis=1)
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: Why don't you just use `"keep_dims": False`?

File: tests/test_TFNetworkLayer.py [hunk 8]
```diff
@@ -3065,6 +3065,71 @@ def mask_func(source, **kwargs):
...
+    ref = ref.mean(axis=0, keepdims=True)
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: This is not correct. This is the mean of the mean. Which is not the same as the mean over all frames directly. I wonder why you test still passes. Maybe `rtol=1e-5` is too large? Check that.

File: tests/test_TFNetworkLayer.py (hunk 7 above)
Comment (vieting) [reply tests/test_TFNetworkLayer.py]: I had it in the network config with `keep_dims` before, probably it doesn't really matter. I changed it anyway.

File: tests/test_TFNetworkLayer.py (hunk 8 above)
Comment (vieting) [reply tests/test_TFNetworkLayer.py]: You are right. I guess then it is easier to just keep the flattening for cases where `axes` contains the time dimension plus other dimensions. What do you think?

Regarding the tolerance, `rtol=1e-6` still passes, `1e-7` does not. I took `1e-5` from the other usage of `assert_allclose()` in this file. Not sure what a sensible threshold is. Anyway, it should not be related to calculating the mean of the mean because currently that is also what is done in the layer.

File: tests/test_TFNetworkLayer.py [hunk 9]
```diff
@@ -3065,6 +3065,63 @@ def mask_func(source, **kwargs):
...
+    max_seq_len = 10
+    feed = make_feed_dict(net.extern_data.data.values(), n_batch=n_batch, n_time=max_seq_len, same_time=True)
+    v = session.run(out, feed_dict=feed)
+    input_len = feed[net.extern_data.data["data"].size_placeholder[0]]
+    input_data = feed[net.extern_data.data["data"].placeholder]
+
+    ref = numpy.zeros([n_batch, n_out])
+    for batch, seq_len in enumerate(input_len):
+      ref[batch, :] = numpy.mean(input_data[batch, :seq_len, :], axis=0)
+    ref = ref.mean(axis=0)
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: Still the problem as explained earlier.

File: tests/test_TFNetworkLayer.py (hunk 8 above)
Comment (albertz) [reply tests/test_TFNetworkLayer.py]: > You are right. I guess then it is easier to just keep the flattening for cases where `axes` contains the time dimension plus other dimensions. What do you think?

Are you talking about the test, or the layer code?

I actually did not realize, the whole logic in the layer for `tf.reduce_mean` is wrong, as you do it the same way there.

Yes,  maybe the flattening is nicer and simpler. I also thought about numerical stability, which is probably better that way.

> Regarding the tolerance, `rtol=1e-6` still passes, `1e-7` does not. I took `1e-5` from the other usage of `assert_allclose()` in this file. Not sure what a sensible threshold is. Anyway, it should not be related to calculating the mean of the mean because currently that is also what is done in the layer.

I think it's ok. I thought that you do it correct in the layer, but wrong here, so I wondered that it did not catch that. But I see now that you do it just as wrong in the layer.

File: TFNetworkLayer.py [hunk 10]
```diff
@@ -4750,21 +4750,23 @@ def __init__(self, mode, axes=None, axis=None, keep_dims=False, enforce_batch_di
...
-          assert x.batch_dim_axis in axes, "not yet implemented otherwise"
-          axes = [a if (a < x.time_dim_axis) else (a - 1)
-                  for a in axes if a != x.time_dim_axis]
-          x = x.copy_time_flattened()
-          x_ = x.placeholder
+          if f == tf.reduce_mean:
+            seq_len_bc = tf.reshape(
+              x.get_sequence_lengths(),
+              [1 if (i != x.batch_dim_axis) else -1 for i in range(x.batch_ndim)])  # (1,..B..,1)
+            x_ = x_ / tf.cast(seq_len_bc, tf.float32)
```
Comment (albertz) [inline TFNetworkLayer.py]: This should not be done multiple times (if this code is run for multiple axes). And this only really make sense if `axis == x.time_dim_axis`.
I think I really would write specific code for `axes == [x.time_dim_axis]`. Otherwise this is a bit too unreadable.

File: tests/test_TFNetworkLayer.py [hunk 11]
```diff
@@ -3065,6 +3065,63 @@ def mask_func(source, **kwargs):
...
+    max_seq_len = 10
+    feed = make_feed_dict(net.extern_data.data.values(), n_batch=n_batch, n_time=max_seq_len, same_time=True)
+    v = session.run(out, feed_dict=feed)
+    input_len = feed[net.extern_data.data["data"].size_placeholder[0]]
+    input_data = feed[net.extern_data.data["data"].placeholder]
+
+    input_data_masked = numpy.copy(input_data)
+    for batch, seq_len in enumerate(input_len):
+      input_data_masked[batch, seq_len:, :] = numpy.nan
+    ref = numpy.nanmean(input_data_masked, axis=(0, 1))
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py]: Ah, interesting, I did not know there is such a function.

File: TFNetworkLayer.py (hunk 10 above)
Comment (albertz) [reply TFNetworkLayer.py]: It's still not so nice. If you read only this code, and not the check a couple of lines above (`axes == [x.time_dim_axis]`), you would think this is a bug. Code should be locally bug-free, as far as possible, e.g. when you remove the check `axes == [x.time_dim_axis]` above, this should still be correct. (This is for easier readability, but also for maintainability, e.g. when changes to the code are made, e.g. someone might think, why do we have this check `axes == [x.time_dim_axis]` there, it should also work without, and then removes it...)
Anyway, I only see the diff here, and don't really know whether other ways to write this are more clear, so let's leave it like that for now.
